core/ipc/sink/common/src/main/java/org/opennms/core/ipc/sink/common/AsyncDispatcherImpl.java

```diff
@@ -0,0 +1,163 @@
+
+public class AsyncDispatcherImpl<W, S extends Message, T extends Message> implements AsyncDispatcher<S>  {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AsyncDispatcherImpl.class);
+
+    private final SyncDispatcher<S> syncDispatcher;
+
+    final RateLimitedLog rateLimittedLogger = RateLimitedLog
+            .withRateLimit(LOG)
+            .maxRate(5).every(Duration.standardSeconds(30))
+            .build();
+
+    final LinkedBlockingQueue<Runnable> queue;
+    final ExecutorService executor;
+
+    public AsyncDispatcherImpl(DispatcherState<W,S,T> state, AsyncPolicy asyncPolicy, SyncDispatcher<S> syncDispatcher) {
+        Objects.requireNonNull(state);
+        Objects.requireNonNull(asyncPolicy);
+        this.syncDispatcher = Objects.requireNonNull(syncDispatcher);
+
+        final RejectedExecutionHandler rejectedExecutionHandler;
+        if (asyncPolicy.isBlockWhenFull()) {
+            queue = new OfferBlockingQueue<>(asyncPolicy.getQueueSize());
+            rejectedExecutionHandler = new ThreadPoolExecutor.AbortPolicy();
+        } else {
+            queue = new LinkedBlockingQueue<Runnable>(asyncPolicy.getQueueSize());
+            // Reject and increase the dropped counter when the queue is full
+            final Counter droppedCounter = state.getMetrics().counter(MetricRegistry.name(state.getModule().getId(), "dropped"));
+            rejectedExecutionHandler = new RejectedExecutionHandler() {
+                @Override
+                public void rejectedExecution(Runnable r, ThreadPoolExecutor e) {
+                    droppedCounter.inc();
+                    throw new RejectedExecutionException("Task " + r.toString() +
+                            " rejected from " +
+                            e.toString());
+                }
+            };
+        }
+
+        state.getMetrics().register(MetricRegistry.name(state.getModule().getId(), "queue-size"), new Gauge<Integer>() {
+            @Override
+            public Integer getValue() {
+                return queue.size();
+            }
+        });
+
+        executor = new ThreadPoolExecutor(
+                asyncPolicy.getNumThreads(),
+                asyncPolicy.getNumThreads(),
+                1000L,
+                TimeUnit.MILLISECONDS,
+                queue,
+                new LogPreservingThreadFactory("OpenNMS.Sink.AsyncDispatcher." + state.getModule().getId(), Integer.MAX_VALUE),
+                rejectedExecutionHandler
+            );
+    }
+
+    /**
+     * Blocks calls to {@link LinkedBlockingQueue#offer(Object)}, converting
+     * these to calls to {@link LinkedBlockingQueue#put(Object)}.
+     *
+     * This is used to block the calling thread when trying to execute
+     * tasks when the queue is already full.
+     *
+     * Make sure that corePoolSize == maximumPoolSize if using
+     * this in an ExecutorService.
+     */
+    private static class OfferBlockingQueue<E> extends LinkedBlockingQueue<E> {
```

Comment:
I've been reading that it's better to use a RejectedExecutionHandler that does the put() instead of changing the semantics of the queue. Then you don't have to worry about what corePoolSize is set to.